command/agent/retry_join.go

```diff
@@ -0,0 +1,90 @@
+package agent
+
+import (
+	"log"
+	"strings"
+	"time"
+)
+
+// DiscoverInterface is an interface for the Discover type in the go-discover
+// library. Using an interface allows for ease of testing.
+type DiscoverInterface interface {
+	// Addrs discovers ip addresses of nodes that match the given filter
+	// criteria.
+	// The config string must have the format 'provider=xxx key=val key=val ...'
+	// where the keys and values are provider specific. The values are URL
+	// encoded.
+	Addrs(string, *log.Logger) ([]string, error)
+
+	// Help describes the format of the configuration string for address
+	// discovery and the various provider specific options.
+	Help() string
+
+	// Names returns the names of the configured providers.
+	Names() []string
+}
+
+// retryJoiner is used to handle retrying a join until it succeeds or all of
+// its tries are exhausted.
+type retryJoiner struct {
+	// join adds the specified servers to the serf cluster
+	join func([]string) (int, error)
+
+	// discover is of type Discover, where this is either the go-discover
+	// implementation or a mock used for testing
+	discover DiscoverInterface
+
+	// errCh is used to communicate with the agent when the max retry attempt
+	// limit has been reached
+	errCh chan struct{}
+
+	// logger is the agent logger.
+	logger *log.Logger
+}
+
+// retryJoin is used to handle retrying a join until it succeeds or all retries
+// are exhausted.
+func (r *retryJoiner) RetryJoin(config *Config) {
+	if len(config.Server.RetryJoin) == 0 || !config.Server.Enabled {
+		return
+	}
+
+	attempt := 0
+
+	r.logger.Printf("[INFO] agent: Joining cluster...")
+
+	for {
+		var addrs []string
+
+		for _, addr := range config.Server.RetryJoin {
+			switch {
+			case strings.Contains(addr, "provider"):
+				servers, err := r.discover.Addrs(addr, r.logger)
+				if err != nil {
+					r.logger.Printf("[ERR] agent: Join error %s", err)
+				} else {
+					addrs = append(addrs, servers...)
+				}
+			default:
+				addrs = append(addrs, addr)
+			}
+		}
+
+		n, err := r.join(addrs)
```

Comment:
Do you need to check for if `len(addrs) == 0` before doing this? If `discover.Addrs(...)` fails above `addrs` may be empty.

Reply:
If we do that check @schmichael suggests then we also need to make sure it retries until `MaxRetryAttempts` rather than return and fail. This is so that if the other peers are still restarting/coming up go discover may not find anything the first time but will find it on a retry.